Day2 puzzle (#20)
scalacenter/scala-advent-of-code: docs/puzzles/day2.md

---
sidebar_position: 2
---

# Day 2: Dive!
by [@mlachkar](https://twitter.com/meriamLachkar)

## Puzzle description

https://adventofcode.com/2021/day/2

## Solution of Part 1

### Parsing the file

> The first step is to model the problem correctly and then parse the input file.

Reviewer: ```suggestion
The first step is to model the problem and to parse the input file.
```

The command can be either `Forward`, `Down` or `Up`. I use an Enumeration to model it. 

```scala
enum Command:
  case Forward(x: Int)
  case Down(x: Int)
  case Up(x: Int)

object Command:
  def from(s: String): Command =
    s match
      case s"forward $x" if x.toIntOption.isDefined => Forward(x.toInt)
      case s"up $x"      if x.toIntOption.isDefined => Up(x.toInt)
      case s"down $x"    if x.toIntOption.isDefined => Down(x.toInt)
      case _ => throw new Exception(s"value $s is not valid command")

  def parseFile(): Iterator[Command] = 
    val input = util.Using.resource(Source.fromFile("input/day2"))(_.mkString)

>     input.linesIterator.map(Command.from)

Reviewer: Not sure it adds anything here

  val input = util.Using.resource(Source.fromFile("input/day2"))(_.mkString)
  val entries: Iterator[Command] = input.linesIterator.map(Command.from)
  val answer = part2(input)

case class PositionWithAim(horizontal: Int, depth: Int, aim: Int):
  def move(p: Command): PositionWithAim =
    p match
      case Command.Forward(x) => PositionWithAim(horizontal + x, depth + x * aim, aim)
      case Command.Down(x)    => PositionWithAim(horizontal, depth, aim + x)
      case Command.Up(x)      => PositionWithAim(horizontal, depth, aim - x)

  def result = horizontal * depth

enum Command:
  case Forward(x: Int)
  case Down(x: Int)
  case Up(x: Int)

object Command:
  def from(s: String): Command =
    s match
      case s"forward $x" if x.toIntOption.isDefined => Forward(x.toInt)
      case s"up $x"      if x.toIntOption.isDefined => Up(x.toInt)
      case s"down $x"    if x.toIntOption.isDefined => Down(x.toInt)
      case _ => throw new Exception(s"value $s is not valid command")

```

## Enum in Scala 3
An enumeration is used to define a type consisting of a set of named values.

We used to used `sealed trait` with `case class` and `case object` to model this set of named values.

> The syntax is more consise and easier to explain than the combination we had in Scala 2. 

Reviewer: Scala 3 enums are more concise and easier to read that the Scala 2 ADTs.

```
enum Command:
  case Forward(x: Int)
  case Down(x: Int)
  case Up(x: Int)
```

is almost the same as:
```
sealed trait Command
object Command {
  case class Forward(x: Int) extends Command
  case class Down(x: Int) extends Command
  case class Up(x: Int) extends Command
}
```

Read [the official documentation](https://docs.scala-lang.org/scala3/reference/enums/enums.html) 
for more details.

## FoldLeft 

> `foldLeft` is a method from the standard library on `IterableOnce` (ie: Seq, List, Iterator, ...).

Reviewer: ```suggestion
`foldLeft` is a method from the standard library on iterable collections (ie: `Seq`, `List`, `Iterator`, ...).
```

It's a super convenient method that allows to iterate from left to right on a list. 

Let's see first an example:

> ```scala
> // signature of foldLeft
> def foldLeft[B](initialElement: B)(op: (B, A) => B): B
> 
> // List sum
> val nums = List(1, 3, 2, 4)
> // equivalent to (((0 + 1) + 3 ) + 2 ) + 4 
> nums.foldLeft(0)((acc, cur) => acc + cur) // 10
> 
> ```

Reviewer: ```suggestion
The signature of foldLeft is:
`scala
def foldLeft[B](initialElement: B)(op: (B, A) => B): B
`

`scala
List(1, 3, 2, 4).foldLeft(0)((acc, cur) => acc + cur) // 10
`

is the same as:
`scala
(((0 + 1) + 3) + 2) + 4
`
```
